### PR TITLE
refactor: move add document logic to use case

### PR DESCRIPTION
**Pull request by eoaksnes**
## What does this pull request change?

* Moves the add document from the document service to the `add_document_use case`.

## Why is this pull request needed?

* Document service is too big, and getting an overview of the code is therefore more complicated than it should. Isolate all document add functionality to the add document feature (use case). 

> NOTE: this is the first PR of several that will refactor the document service.

### CODE REVIEW

File: src/features/document/use_cases/add_document_use_case.py [hunk 1]
```diff
@@ -1,28 +1,220 @@
...
-    address: str,
+    address: Address,
+    document_service: DocumentService,
     files: Optional[List[UploadFile]] = None,
     update_uncontained: Optional[bool] = False,
-    repository_provider=get_data_source,
-):
-    document_service = DocumentService(repository_provider=repository_provider, user=user)
+) -> dict:
+    """Add document to a data source or existing entity.
```
Comment (KristianKjerstad) [inline src/features/document/use_cases/add_document_use_case.py:197]: the use case is also used to add items to list
```suggestion
    """Add document to a data source or existing entity. Can also be used to add (complex) items to a list. 
```

File: src/features/document/use_cases/add_document_use_case.py [hunk 2]
```diff
@@ -1,28 +1,220 @@
...
+
+    Returns:
+        A dict that contains the ID of the added document.
+    """
+    validate_entity_against_self(document, document_service.get_blueprint)
+
+    if not address.path:
+        return _add_document_to_data_source(address.data_source, document, update_uncontained, document_service)
+
+    return _add_document_to_entity(
```
Comment (KristianKjerstad) [inline src/features/document/use_cases/add_document_use_case.py:214]:
```suggestion
    return _add_document_to_entity_or_list(
```

File: src/features/document/use_cases/add_document_use_case.py [hunk 3]
```diff
@@ -1,28 +1,220 @@
...
-):
-    document_service = DocumentService(repository_provider=repository_provider, user=user)
+) -> dict:
+    """Add document to a data source or existing entity.
 
-    document = document_service.add(
-        address=Address.from_absolute(address),
+    Args:
+        document: The entity to be added
+        address: Reference to a package or a data source
```
Comment (KristianKjerstad) [inline src/features/document/use_cases/add_document_use_case.py:201]:
```suggestion
        address: Reference to a package, attribute inside an entity (either a list or a complex attribute) or a data source
```

File: src/features/document/use_cases/add_document_use_case.py [hunk 4]
```diff
@@ -1,28 +1,220 @@
...
+    address: Address,
+    document: dict,
+    files: dict[str, BinaryIO] | None,
+    update_uncontained: Optional[bool],
+    document_service: DocumentService,
+) -> dict:
+    """Add the document to an existing entity.
+
+    Args:
+       address: Reference to an existing entity
```
Comment (KristianKjerstad) [inline src/features/document/use_cases/add_document_use_case.py:71]:
```suggestion
       address: Reference to an existing entity or to an attribute (complex or list attribute) inside an entity.
```

**Review by KristianKjerstad — COMMENTED** [review]
For this structure to make sense it looks like we have to remove the entire document service and move all that logic into use cases. Is that the plan?

**Comment by eoaksnes** [comment]
> For this structure to make sense it looks like we have to remove the entire document service and move all that logic into use cases. Is that the plan?

Not everything will be removed, the get and save will still be part of the document service (or in the shape of another component). Since all CRUD  operations will be using get and save.

**Review by pbullhove — APPROVED** [review]

**Pull request merged**